### DISCUSSION TITLE
Custom request headers

### DISCUSSION THREAD

**Discussion by vetalstar** [vercel/vercel#6555] [Help]
Hi!

Is there any way to set custom request header like X-Vercel-IP-City?
I would like to independently detect the user's city by myself and than set the x-app-city header. This will allow to do redirects between subdomains in next js app.

**Comment by leerob** [comment]
Yes, you can do:

```js
res.setHeader('x-app-city', yourCustomValue)
```

**Reply by vetalstar** [reply]
Hi @leerob. I think you didn't understand me.
I've found the example of redirecting user based on their location:
https://nextjs.org/blog/next-10-2#routing-based-on-headers-and-query-string-parameters
```
// next.config.js

module.exports = {
  async redirects() {
    return [
      {
        source: '/:path((?!uk/).*)',
        has: [
          {
            type: 'header',
            key: 'x-vercel-ip-country',
            value: 'GB'
          }
        ],
        destination: '/uk/:path*',
        permanent: true
      }
    ]
  }
}
```

For example:
The user is from New York. I want to use my geolocation provider and set request header x-user-city: NY.
And then configure server redirect (for SEO) to ny.example.com with such value of the x-user-city header in the next.config.js.

Is there any way to save SSG and determine the user's location with custom provider and set request headers with vercel?

**Reply by leerob** [reply]
- If you need to look at the incoming request, you can't use `getStaticProps`. You can use `getServerSideProps` plus edge caching, by setting request headers.
- If you want geolocation, this is built into Vercel https://vercel.com/changelog/ip-geolocation-for-serverless-functions

**Reply by vetalstar** [reply]
@leerob Can I handle requests and modify them in vercel before next js static page? I think I search in vercel something like this https://developers.cloudflare.com/workers/examples/country-code-redirect
as a way to set custom request header for redirecting between city hosts

**Reply by leerob** [reply]
Yes, that's the code snippet above with redirects. You can use that in conjunction with static pages!

```js
// next.config.js

module.exports = {
  async redirects() {
    return [
      {
        source: '/',
        has: [
          {
            type: 'header',
            key: 'x-vercel-ip-country',
            value: 'GB'
          }
        ],
        destination: '/uk',
        permanent: false
      }
    ]
  }
}
```

Where `/uk` could be a static page.

**Reply by vetalstar** [reply]
I understand how to manage redirects in the next js app.
I want to have a custom geo detection instead of x-vercel-ip-country and I'll need to set x-custom-ip-city to the request with my value. Can I do it in vercel as such as in cloudfare?
